### PR TITLE
modules/SceAppMgrUser: Stub some function with str(n)cpy.

### PR DESCRIPTION
**Pull request by Zangetsu38**

### CODE REVIEW

File: vita3k/modules/SceDriverUser/SceAppMgrUser.cpp [hunk 1]
```diff
@@ -172,9 +181,17 @@ EXPORT(int, sceAppMgrForceUmount) {
     return UNIMPLEMENTED();
 }
 
-EXPORT(int, sceAppMgrGameDataMount) {
-    TRACY_FUNC(sceAppMgrGameDataMount);
-    return UNIMPLEMENTED();
+EXPORT(int, sceAppMgrGameDataMount, Ptr<const char> app_path, Ptr<const char> patch_path, Ptr<const char> rif_path, char *mount_point) {
```
Comment (Macdu) [inline vita3k/modules/SceDriverUser/SceAppMgrUser.cpp:184]:
```suggestion
EXPORT(int, sceAppMgrGameDataMount, const char* app_path, const char* patch_path, const char* rif_path, char *mount_point) {
```

Comment (Zangetsu38) [reply vita3k/modules/SceDriverUser/SceAppMgrUser.cpp:184]: is sad but no works with this, get crash

Comment (Macdu) [reply vita3k/modules/SceDriverUser/SceAppMgrUser.cpp:184]: Can you show what the crash is? there is really no reason for this to cause a crash

File: vita3k/modules/SceDriverUser/SceAppMgrUser.cpp [hunk 2]
```diff
@@ -289,9 +306,11 @@ EXPORT(int, sceAppMgrGetPidListForShell) {
     return UNIMPLEMENTED();
 }
 
-EXPORT(int, sceAppMgrGetRawPath) {
+EXPORT(int, sceAppMgrGetRawPath, char *path, char *resolved_path, int resolved_path_size) {
     TRACY_FUNC(sceAppMgrGetRawPath);
```
Comment (bookmist) [inline vita3k/modules/SceDriverUser/SceAppMgrUser.cpp:310]:
```suggestion
    TRACY_FUNC(sceAppMgrGetRawPath,path,resolved_path,resolved_path_size);
```

Comment (Zangetsu38) [reply vita3k/modules/SceDriverUser/SceAppMgrUser.cpp:310]: oupsy

File: vita3k/modules/SceDriverUser/SceAppMgrUser.cpp (hunk 1 above)
Comment (Zangetsu38) [reply vita3k/modules/SceDriverUser/SceAppMgrUser.cpp:184]: yes you right, i have no understund here...

Comment (Zangetsu38) [reply vita3k/modules/SceDriverUser/SceAppMgrUser.cpp:184]: ah no ok i remember where
image
i have no idea if i no using Ptr <> get that and crash

Comment (Zangetsu38) [reply vita3k/modules/SceDriverUser/SceAppMgrUser.cpp:184]: ok sorry, stupid error caused by other things strange

Comment (Zangetsu38) [reply vita3k/modules/SceDriverUser/SceAppMgrUser.cpp]: done so

File: vita3k/modules/SceDriverUser/SceAppMgrUser.cpp (hunk 2 above)
Comment (Zangetsu38) [reply vita3k/modules/SceDriverUser/SceAppMgrUser.cpp]: done

File: vita3k/modules/SceDriverUser/SceAppMgrUser.cpp [hunk 3]
```diff
@@ -75,8 +75,17 @@ EXPORT(int, sceAppMgrAddContMount) {
     return UNIMPLEMENTED();
 }
 
-EXPORT(int, sceAppMgrAppDataMount) {
-    TRACY_FUNC(sceAppMgrAppDataMount);
+EXPORT(int, sceAppMgrAppDataMount, int id, const char *mount_point) {
+    TRACY_FUNC(sceAppMgrAppDataMount, id, mount_point);
+    switch (id) {
+    case 105:
```
Comment (Macdu) [inline vita3k/modules/SceDriverUser/SceAppMgrUser.cpp:81]: Isn't there an enum for all the possible IDs?

File: vita3k/modules/SceDriverUser/SceAppMgrUser.cpp [hunk 4]
```diff
@@ -172,9 +181,17 @@ EXPORT(int, sceAppMgrForceUmount) {
...
-    return UNIMPLEMENTED();
+EXPORT(int, sceAppMgrGameDataMount, const char *app_path, const char *patch_path, const char *rif_path, char *mount_point) {
+    TRACY_FUNC(sceAppMgrGameDataMount, app_path, patch_path, rif_path, mount_point);
+    STUBBED("Using strcpy");
+    if (!std::string(app_path).empty())
+        strcpy(mount_point, fmt::format("{}/", app_path).c_str());
+    else if (!std::string(patch_path).empty())
+        strcpy(mount_point, fmt::format("{}/", patch_path).c_str());
+    else if (!std::string(rif_path).empty())
+        strcpy(mount_point, fmt::format("{}/", rif_path).c_str());
```
Comment (Macdu) [inline vita3k/modules/SceDriverUser/SceAppMgrUser.cpp:192]:
```suggestion
    if (strlen(app_path) > 0)
        strcpy(mount_point, fmt::format("{}/", app_path).c_str());
    else if (strlen(patch_path) > 0)
        strcpy(mount_point, fmt::format("{}/", patch_path).c_str());
    else if (strlen(rif_path) > 0)
        strcpy(mount_point, fmt::format("{}/", rif_path).c_str());
```

File: vita3k/modules/SceDriverUser/SceAppMgrUser.cpp (hunk 3 above)
Comment (Zangetsu38) [reply vita3k/modules/SceDriverUser/SceAppMgrUser.cpp:81]: off course, i have just only tested it

File: vita3k/modules/SceDriverUser/SceAppMgrUser.cpp (hunk 4 above)
Comment (Zangetsu38) [reply vita3k/modules/SceDriverUser/SceAppMgrUser.cpp]: done

File: vita3k/modules/SceDriverUser/SceAppMgrUser.cpp (hunk 3 above)
Comment (Zangetsu38) [reply vita3k/modules/SceDriverUser/SceAppMgrUser.cpp]: done, and reworked, is wrong before

File: vita3k/modules/SceDriverUser/SceAppMgrUser.cpp [hunk 5]
```diff
@@ -75,9 +75,37 @@ EXPORT(int, sceAppMgrAddContMount) {
     return UNIMPLEMENTED();
 }
 
-EXPORT(int, sceAppMgrAppDataMount) {
-    TRACY_FUNC(sceAppMgrAppDataMount);
-    return UNIMPLEMENTED();
+EXPORT(int, sceAppMgrAppDataMount, int id, const char *mount_point) {
```
[the remark this answers is not in the document]
Comment (Zangetsu38) [reply vita3k/modules/SceDriverUser/SceAppMgrUser.cpp:78]: true lol

[the remark this answers is not in the document]
Comment (Zangetsu38) [reply vita3k/modules/SceDriverUser/SceAppMgrUser.cpp]: done

File: vita3k/modules/SceDriverUser/SceAppMgrUser.cpp [hunk 6]
```diff
@@ -172,9 +200,17 @@ EXPORT(int, sceAppMgrForceUmount) {
...
 }
 
-EXPORT(int, sceAppMgrGameDataMount) {
-    TRACY_FUNC(sceAppMgrGameDataMount);
-    return UNIMPLEMENTED();
+EXPORT(int, sceAppMgrGameDataMount, const char *app_path, const char *patch_path, const char *rif_path, char *mount_point) {
+    TRACY_FUNC(sceAppMgrGameDataMount, app_path, patch_path, rif_path, mount_point);
+    STUBBED("Using strcpy");
+    if (strlen(app_path) > 0)
+        strcpy(mount_point, fmt::format("{}/", app_path).c_str());
```
Comment (Macdu) [inline vita3k/modules/SceDriverUser/SceAppMgrUser.cpp:207]:
```suggestion
        fmt::format_to(mount_point, "{}/", app_path);
```
And same for all the other times this is used, this looks much cleaner.

Comment (Zangetsu38) [reply vita3k/modules/SceDriverUser/SceAppMgrUser.cpp:207]: oh, i no know this

File: vita3k/modules/SceDriverUser/SceAppMgrUser.cpp [hunk 7]
```diff
@@ -344,9 +382,12 @@ EXPORT(int, sceAppMgrGetSystemDataFilePlayReady) {
...
 }
 
-EXPORT(int, sceAppMgrGetUserDirPath) {
-    TRACY_FUNC(sceAppMgrGetUserDirPath);
-    return UNIMPLEMENTED();
+EXPORT(int, sceAppMgrGetUserDirPath, int partition_id, char *userDirPath, SceSize path_maxlen) {
+    TRACY_FUNC(sceAppMgrGetUserDirPath, partition_id, userDirPath, path_maxlen);
+    STUBBED("Using strncpy");
+    const auto partition = partition_id == 1 ? "ur0:" : "ux0:";
+    strncpy(userDirPath, fmt::format("{}user/{}", partition, emuenv.io.user_id).c_str(), path_maxlen);
```
Comment (Macdu) [inline vita3k/modules/SceDriverUser/SceAppMgrUser.cpp:389]:
```suggestion
    strncpy(userDirPath, fmt::format("{}user/{}", partition, emuenv.io.user_id).c_str(), path_maxlen);
```
Use fmt::format_to_n

Comment (Zangetsu38) [reply vita3k/modules/SceDriverUser/SceAppMgrUser.cpp:389]: done oki

**Review by Macdu — APPROVED** [review]

**Pull request merged**